gateway/request_context.go

```diff
@@ -281,32 +282,102 @@ func (rc *requestContext) BuildProtocolContextFromHTTP(httpReq *http.Request) er
 // See the following link for more details:
 // https://en.wikipedia.org/wiki/Template_method_pattern
 func (rc *requestContext) HandleRelayRequest() error {
+	triedEndpoints := []protocol.EndpointAddr{}
+	return rc.handleRelayRequestWithRetries(maxRequestRetriesDefault, 0, triedEndpoints)
+}
+
+// handleRelayRequestWithRetriesRecursive is the recursive implementation of HandleRelayRequestWithRetries.
+func (rc *requestContext) handleRelayRequestWithRetries(
+	maxRetries,
+	currentAttempt int,
+	triedEndpoints []protocol.EndpointAddr,
+) error {
 	// Send the service request payload, through the protocol context, to the selected endpoint.
 	endpointResponse, err := rc.protocolCtx.HandleServiceRequest(rc.qosCtx.GetServicePayload())
-	if err != nil {
-		rc.logger.Warn().Err(err).Msg("Failed to send a relay request.")
-		// TODO_TECHDEBT(@commoddity): the correct reaction to a failure in sending the relay to an endpoint and getting
-		// a response could be retrying with another endpoint, depending on the error.
-		// This should be revisited once a retry mechanism for failed relays is within scope.
-		//
-		// TODO_TECHDEBT(@adshmh): use the relay error in the response returned to the user.
+	if err == nil || currentAttempt >= maxRetries {
+		// Success! Process the response and return.
+		rc.qosCtx.UpdateWithResponse(endpointResponse.EndpointAddr, endpointResponse.Bytes)
+		return nil
+	}
+
+	// Log the failure
+	rc.logger.Warn().Err(err).Msgf("Failed to send a relay request on attempt %d of %d.", currentAttempt+1, maxRetries)
+
+	// Track the failed endpoint if we have response information
+	if endpointResponse.EndpointAddr != "" {
+		triedEndpoints = append(triedEndpoints, endpointResponse.EndpointAddr)
+	}
+
+	// Try to get a different endpoint and rebuild the protocol context
+	if retryErr := rc.selectNewEndpointForRetry(triedEndpoints); retryErr != nil {
```

Comment:
This looks redundant: The previously selected endpoint should already be sanctioned at the Protocol level if it returned an invalid response or failed in any way.